File: 0-frameworks-in-backend/backend/2-csharp/Cat/CatService.cs
// The cat "module" — a dedicated namespace as the bounded context.
namespace FrameworksDemo.Cat;

/// <summary>
/// Immutable data record representing a single cat.
/// <para>
/// <c>System.Text.Json</c> serializes properties in camelCase by default:
/// <c>{"id": .., "name": ..}</c>, matching the API contract.
/// </para>
/// </summary>
/// <param name="Id">Unique identifier of the cat.</param>
/// <param name="Name">Display name of the cat.</param>
public record Cat(int Id, string Name);

/// <summary>
/// Business-logic service for cat data.
/// <para>
/// Registered as a singleton via <c>AddSingleton&lt;CatService&gt;()</c> in Program.cs.
/// The ASP.NET IoC container creates one instance and injects it wherever required
/// (Minimal API handlers and <c>DogService</c> constructor).
/// </para>
/// </summary>
public class CatService
{
    /// <summary>Shared, immutable cat dataset used by all public methods.</summary>
    private readonly List<Cat> _cats = new() { new(1, "Milo"), new(2, "Luna") };

    /// <summary>
    /// Return the full cat list — consumed by <c>GET /cats</c>.
    /// </summary>
    /// <returns>Read-only list of <see cref="Cat"/> records.</returns>
    public IReadOnlyList<Cat> All() => _cats;

    /// <summary>
    /// Provide a status hint consumed by <see cref="FrameworksDemo.Dog.DogService"/> via DI.
    /// <para>
    /// Exists solely to prove that <c>DogService</c> can call methods on the injected
    /// <c>CatService</c> without creating a new instance.
    /// </para>
    /// </summary>
    /// <returns>Fixed status string <c>"cat-network-ready"</c>.</returns>
    public string SpyHint() => "cat-network-ready";
}

File: 0-frameworks-in-backend/backend/1-java/src/main/java/com/example/cat/CatService.java
package com.example.cat;

import java.util.List;
import org.springframework.stereotype.Service;

/**
 * Business-logic service for cat data.
 *
 * <p>{@code @Service} registers this class as a singleton bean in Spring's IoC
 * container — equivalent to NestJS's {@code @Injectable()}. The container creates
 * exactly one instance and injects it wherever required (e.g., into
 * {@link CatController} and {@code DogService}).
 */
@Service
public class CatService {

    /**
     * Shared, immutable cat dataset used by all public methods.
     * Static data keeps the lesson focused on DI rather than database setup.
     */
    private final List<Cat> cats = List.of(new Cat(1, "Milo"), new Cat(2, "Luna"));

    /**
     * Return the full cat list — consumed by {@code GET /cats}.
     *
     * @return immutable list of {@link Cat} records
     */
    public List<Cat> all() {
        // Return the in-memory list; no DB call needed for this DI demo.
        return cats;
    }

    /**
     * Provide a status hint string consumed by {@code DogService} via cross-module DI.
     *
     * <p>This method exists solely to prove that {@code DogService} can call methods
     * on the injected {@link CatService} without creating a new instance.
     *
     * @return fixed status string {@code "cat-network-ready"}
     */
    public String spyHint() {
        return "cat-network-ready";
    }
}
